### ISSUE TITLE
decay, decay_rate and decay_steps not implemented

### ISSUE THREAD

**Issue by carlogrisetti** [uber/ludwig#839]
Using latest master it seems to me that `decay`, `decay_rate `and `decay_steps `are not affecting the learning rate at all. Looking in the trainer model, they don't even seem to be used in the train function.
https://github.com/uber/ludwig/blob/62430e4a0dd7a4fda08d6dcd615fbdbbf53c5377/ludwig/models/trainer.py#L166-L195

`learning_rate `and `learning_rate_warmup_epochs` instead work fine (and I see them parsed in the train function)
Am I missing something?
Maybe it's related to the TF2 porting?

**Comment by w4nderlust** [comment]
#898 solves this. Let me know if it's not ok.

**Comment by carlogrisetti** [comment]
I am finally able to test decay and decay_rate, but it seems that the learning rate reduction due to decay is only applied one time, and then never again

ie: I set LR to 0.1, decay to true, decay_rate to 0.9 and LR goes down to 0.09986 after the first epoch and then stays that way forever

Can you check this? Do you prefer to have a new issue opened?

Thanks

**Comment by w4nderlust** [comment]
So this is kinda weird. inspecting this I figured out some issues with the way the learning rate is computed. I'm fixing them, but yours is different. Also, changing the reporting of the learning rate per batch, as the decay (and the warm up) ar applead on a per batch basis.
Will probably put up a PR soonfor you to try.

**Comment by w4nderlust** [comment]
Allright, this is the new behavior after the fixes.

Fixes the place where the lr computation is done and some parameters.

Now this happens without warmup and decay:
![nothing](https://user-images.githubusercontent.com/349256/97513485-a961d480-1949-11eb-8bfc-ac898d6cb6d9.png)

This when warmup is on
![warmup](https://user-images.githubusercontent.com/349256/97513499-b1217900-1949-11eb-8be1-b7334eb37f2f.png)

This when decay is on
![decay](https://user-images.githubusercontent.com/349256/97513520-bb437780-1949-11eb-900f-25121bffaee7.png)

This when decay with staircase is on
![decay_staircase](https://user-images.githubusercontent.com/349256/97513534-c4ccdf80-1949-11eb-9df0-cd1aa86ee103.png)

And these wen warmup and decay are on at the same time
![warmup_decay](https://user-images.githubusercontent.com/349256/97513554-cf877480-1949-11eb-8be3-296fd1edafad.png)
![warmup_decay_staircase](https://user-images.githubusercontent.com/349256/97513558-d1e9ce80-1949-11eb-9ad8-9ad4ee2536b7.png)

Looks reasonable to me :) can you doublecheck with your usecase please?

**Comment by carlogrisetti** [comment]
That looks good from your comment, will try and see today.
Warmup was already okay, I don't understand if you found some more issues there or if was just an FYI mentioning warmup also.

There's only the first graph that is suspiciously aligned with 0, and I noted that it happens when the LR is waaaay too low, and maybe there are some precision/rounding mechanisms that set it to zero in the graphs (I think we can more than live with it, since it only applies to the first few epochs in a warmup use case, and never happened in the latter part of the training, since it should have gone down to non usable LR values).

Will report back :)

**Comment by carlogrisetti** [comment]
I had a weird feedback, like that warmup was always set to 10, but this is due to the fact that i have mistaken step_learning_rate for epoch_learning_rate.

It is all correct and working fine!

**Comment by w4nderlust** [comment]
In the first graph, I guess the range set by the tensorboard doesn't allow to see it was actually a 0.01 learning rate, but i manually checked, and the 0.01 just stayed constant throughout.

Glad this worked, merging the PR and closing.
